cis194/week1/zhansongl/credit_card.hs

```diff
@@ -0,0 +1,125 @@
+import Test.QuickCheck
+import Test.Hspec
+
+toDigitsRev :: Integer -> [Integer]
+toDigitsRev n
+  | n >= 0 = digit n
+  | otherwise = []
+  where digit 0 = []
+        digit n = n `mod` 10 : digit (n `div` 10)
+
+toDigitsRev_prop n = (reverse . joinDigits $ toDigitsRev n) == (showDigits n)
+  where types = n::Integer
+        joinDigits xs = foldr (++) "" $ map show xs
+        showDigits n
+          | n <= 0 = "" -- weird restriction, but okay
+          | otherwise = show n
+
+toDigits :: Integer -> [Integer]
+toDigits = reverse . toDigitsRev
+
+toDigits_prop n = (joinDigits $ toDigits n) == (showDigits n)
+  where types = n::Integer
+        joinDigits xs = foldr (++) "" $ map show xs
+        showDigits n
+          | n <= 0 = "" -- weird restriction, but okay
+          | otherwise = show n
+
+doubleEveryOtherRev :: [Integer] -> [Integer]
+doubleEveryOtherRev [] = []
+doubleEveryOtherRev [x] = [x]
+doubleEveryOtherRev (x:y:xs) = (x:2*y:doubleEveryOtherRev xs)
+
+doubleEveryOtherRev_prop1 xs = (2 * (sum xs)) == sum (doubleEveryOtherRev $ padList xs)
+  where types = xs::[Integer]
+        padList [] = []
+        padList [x] = [0,x]
+        padList (x:xs) = (0:x:padList xs)
+
+doubleEveryOtherRev_prop2 xs = (sum xs) == sum (doubleEveryOtherRev $ padList xs)
+  where types = xs::[Integer]
+        padList [] = []
+        padList [x] = [x,0]
+        padList (x:xs) = (x:0:padList xs)
+
+doubleEveryOtherRev_prop3 xs = (sum xs) == sum (doubleEveryOtherRev $ padList xs)
+  where types = xs::[Integer]
+        padList [] = []
+        padList [x] = [x]
+        padList (x:xs) = (x:0:padList xs)
+
+doubleEveryOther :: [Integer] -> [Integer]
+doubleEveryOther = reverse . doubleEveryOtherRev . reverse
```

Comment:
Given we are going to sum the digits after double every other digits from the right, do we need to reverse the list back?

Reply:
Hmm, if the only goal is to finish `validate`, then I suppose there is no point in reversing the list back. For the sake of this exercise though, if the list is returned in reverse it doesn't satisfy the requirements.